In [ ]:
import numpy as np

In [ ]:
def evalf(x,n):
  assert type(x) is np.ndarray and len(x) == n

  sum = 0

  for i in range(1,n+1):
    sum += ((x[i-1]-2**i)**2)/(8**i)
  return sum

In [ ]:
evalf(np.array([1,1]),2)

0.265625

In [ ]:
def evalg(x,n):
  assert type(x) is np.ndarray and len(x) == n

  arr = np.array([])

  for i in range(1,n+1):
    j = 2*(x[i-1]-2**i)/(8**i)
    arr = np.concatenate((arr,np.array([j])),axis=0)
  return arr

In [ ]:
evalg(np.array([1,1]),2)

array([-0.25   , -0.09375])

In [ ]:
def compute_steplength_exact(gradf, A, n):
  assert type(gradf) is np.ndarray and len(gradf) ==  n
  assert type(A) is np.ndarray and A.shape[0] == n and  A.shape[1] == n

  t1 = np.matmul(gradf,gradf)/2
  t2 = np.matmul(np.matmul(A,gradf),gradf)

  step_length = t1/t2
  
  return step_length

In [ ]:
def compute_steplength_backtracking(x, gradf, alpha_start, rho, gamma, n):
  assert type(x) is np.ndarray and len(x) == n
  assert type(gradf) is np.ndarray and len(gradf) == n
  
  alpha = alpha_start

  while evalf(x+alpha*(-gradf),n) > evalf(x,n) + gamma*alpha*np.matmul(gradf.transpose(),-gradf):
    alpha = rho*alpha

  return alpha

In [ ]:
EXACT_LINE_SEARCH = 1
BACKTRACKING_LINE_SEARCH = 2
CONSTANT_STEP_LENGTH = 3

In [ ]:
def find_minimizer(start_x, tol, line_search_type, n, *args):
  assert type(start_x) is np.ndarray and len(start_x) == n
  assert type(tol) is float and tol>=0
  A = np.array([[1/8, 0,0],[0,1/64,0],[0,0,1/512]])
  x = start_x
  g_x = evalg(x,n)

  #initialization for backtracking line search
  if(line_search_type == BACKTRACKING_LINE_SEARCH):
    alpha_start = args[0]
    rho = args[1]
    gamma = args[2]
    #print('Params for Backtracking LS: alpha start:', alpha_start, 'rho:', rho,' gamma:', gamma)

  k = 0
  #print('iter:',k, ' x:', x, ' f(x):', evalf(x,n), ' grad at x:', g_x, ' gradient norm:', np.linalg.norm(g_x))

  while (np.linalg.norm(g_x) > tol): #continue as long as the norm of gradient is not close to zero upto a tolerance tol
  
    if line_search_type == EXACT_LINE_SEARCH:
      step_length = compute_steplength_exact(g_x, A, n) #call the new function you wrote to compute the steplength
      #raise ValueError('EXACT LINE SEARCH NOT YET IMPLEMENTED')
    elif line_search_type == BACKTRACKING_LINE_SEARCH:
      step_length = compute_steplength_backtracking(x,g_x, alpha_start,rho, gamma, n) #call the new function you wrote to compute the steplength
      #raise ValueError('BACKTRACKING LINE SEARCH NOT YET IMPLEMENTED')
    elif line_search_type == CONSTANT_STEP_LENGTH: #do a gradient descent with constant step length
      step_length = 0.1
    else:  
      raise ValueError('ReEnter Line Search Type')
    
    #implement the gradient descent steps here   
    x = np.subtract(x, np.multiply(step_length,g_x)) #update x = x - step_length*g_x
    k += 1 #increment iteration
    g_x = evalg(x,n) #compute gradient at new point

    #print('iter:',k, ' x:', x, ' f(x):', evalf(x,n), ' grad at x:', g_x, ' gradient norm:', np.linalg.norm(g_x))
  return x,evalf(x,n),k


In [ ]:
my_start_x = np.array([10,10,10])
my_tol= 1e-5

In [ ]:
x_opt,f_min,k = find_minimizer(my_start_x, my_tol, CONSTANT_STEP_LENGTH,3)
print('Constant Step Length:','\nOptimizer:',x_opt,'\nMinimum Function Value:',f_min)

Constant Step Length: 
Optimizer: [2.         4.         8.00255956] 
Minimum Function Value: 1.279561039741878e-08


#Ans 2:

Since no data for this question has been provided, I have taken the following values:

start_x = [10,10,10]

tolerance = $10^{-5}$

step length = 0.1

Using Constant Step Length :

Optimizer: [2. , 4. , 8.00255956]

Minimum Function Value: 1.279561039741878e-08

In [ ]:
q3_start_x = np.array([1/64,1/8,1])
q3_tol = 1e-10

In [ ]:
x_opt,f_min,k = find_minimizer(q3_start_x, q3_tol, EXACT_LINE_SEARCH,3)
print('Exact Line Search:','\nOptimizer:',x_opt,'\nMinimum Function Value:',f_min,'\nNo. of iterations:',k)

Exact Line Search: 
Optimizer: [2.         4.         7.99999998] 
Minimum Function Value: 9.150071377581033e-19 
No. of iterations: 269


In [ ]:
x_opt,f_min,k = find_minimizer(q3_start_x, q3_tol, BACKTRACKING_LINE_SEARCH,3,1,0.5,0.5)
print('Backtracking Line Search:','\nOptimizer:',x_opt,'\nMinimum Function Value:',f_min,'\nNo. of iterations:',k)

Backtracking Line Search: 
Optimizer: [2.         4.         7.99999997] 
Minimum Function Value: 1.2748574165464873e-18 
No. of iterations: 4964


#Ans 3:

We can observe that Backtraining Line Search uses around 18 times more iterations than what Exact Line Search uses.

In [ ]:
def find_minimizer(start_x, tol, line_search_type, n, *args):
  assert type(start_x) is np.ndarray and len(start_x) == n
  assert type(tol) is float and tol>=0
  A = np.array([[1/8, 0,0,0],[0,1/64,0,0],[0,0,1/512,0],[0,0,0,1/4096]])
  x = start_x
  g_x = evalg(x,n)

  #initialization for backtracking line search
  if(line_search_type == BACKTRACKING_LINE_SEARCH):
    alpha_start = args[0]
    rho = args[1]
    gamma = args[2]
    #print('Params for Backtracking LS: alpha start:', alpha_start, 'rho:', rho,' gamma:', gamma)

  k = 0
  #print('iter:',k, ' x:', x, ' f(x):', evalf(x,n), ' grad at x:', g_x, ' gradient norm:', np.linalg.norm(g_x))

  while (np.linalg.norm(g_x) > tol): #continue as long as the norm of gradient is not close to zero upto a tolerance tol
  
    if line_search_type == EXACT_LINE_SEARCH:
      step_length = compute_steplength_exact(g_x, A, n) #call the new function you wrote to compute the steplength
      #raise ValueError('EXACT LINE SEARCH NOT YET IMPLEMENTED')
    elif line_search_type == BACKTRACKING_LINE_SEARCH:
      step_length = compute_steplength_backtracking(x,g_x, alpha_start,rho, gamma, n) #call the new function you wrote to compute the steplength
      #raise ValueError('BACKTRACKING LINE SEARCH NOT YET IMPLEMENTED')
    elif line_search_type == CONSTANT_STEP_LENGTH: #do a gradient descent with constant step length
      step_length = 0.1
    else:  
      raise ValueError('ReEnter Line Search Type')
    
    #implement the gradient descent steps here   
    x = np.subtract(x, np.multiply(step_length,g_x)) #update x = x - step_length*g_x
    k += 1 #increment iteration
    g_x = evalg(x,n) #compute gradient at new point

    #print('iter:',k, ' x:', x, ' f(x):', evalf(x,n), ' grad at x:', g_x, ' gradient norm:', np.linalg.norm(g_x))
  return x,evalf(x,n),k


In [ ]:
q4_start_x = np.array([1/512,1/64,1/8,1])
q4_tol = 1e-10

In [ ]:
x_opt,f_min,k = find_minimizer(q4_start_x, q4_tol, EXACT_LINE_SEARCH,4)
print('Exact Line Search:','\nOptimizer:',x_opt,'\nMinimum Function Value:',f_min,'\nNo. of iterations:',k)

Exact Line Search: 
Optimizer: [ 2.          4.          8.         15.99999981] 
Minimum Function Value: 8.8565993523583e-18 
No. of iterations: 2013


In [ ]:
x_opt,f_min,k = find_minimizer(q4_start_x, q4_tol, BACKTRACKING_LINE_SEARCH,4,1,0.5,0.5)
print('Backtracking Line Search:','\nOptimizer:',x_opt,'\nMinimum Function Value:',f_min,'\nNo. of iterations:',k)

Backtracking Line Search: 
Optimizer: [ 2.         4.         8.        15.9999998] 
Minimum Function Value: 1.0237544252113035e-17 
No. of iterations: 37079


#Ans 4:

We can gather similar observations here as in the case for N = 3. Backtracking Line Search uses significantly more iterations (around 18 times more).

#Ans 5:

Considering the case for N = 3 and N = 4, we can say that for N $>$ 4 too Backtracking Line Search takes significantly more no. of iterations than Exact Line Search.